zkvm/src/prover.rs

```diff
@@ -1,25 +1,38 @@
 use bulletproofs::r1cs;
+use bulletproofs::{BulletproofGens, PedersenGens};
 use curve25519_dalek::ristretto::CompressedRistretto;
 use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use std::collections::VecDeque;
 
 use crate::errors::VMError;
 use crate::ops::Instruction;
 use crate::point_ops::PointOp;
+use crate::signature::Signature;
 use crate::types::*;
-use crate::vm::{Delegate, RunTrait};
+use crate::vm::{Delegate, RunTrait, Tx, VM};
 
 pub struct Prover<'a, 'b> {
     signtx_keys: Vec<Scalar>,
     cs: r1cs::Prover<'a, 'b>,
 }
 
-pub struct RunProver {}
+pub struct ProverRun {
+    program: VecDeque<Instruction>,
+}
 
 impl<'a, 'b> Delegate<r1cs::Prover<'a, 'b>> for Prover<'a, 'b> {
-    type RunType = RunProver;
+    type RunType = ProverRun;
 
-    fn commit_variable(&mut self, com: &Commitment) -> (CompressedRistretto, r1cs::Variable) {
-        unimplemented!()
+    fn commit_variable(
+        &mut self,
+        com: &Commitment,
+    ) -> Result<(CompressedRistretto, r1cs::Variable), VMError> {
+        let (v, v_blinding) = match com {
+            Commitment::Open(w) => (w.value.into(), w.blinding),
+            Commitment::Closed(_) => return Err(VMError::WitnessMissing),
+        };
+        Ok(self.cs.commit(v, v_blinding))
     }
 
     fn verify_point_op<F>(&mut self, point_op_fn: F)
```

Comment:
how about making this failable (and closure failable too), so that:

1. Verifier can decide whether to do batching or check and fail early (for diagnostics)
2. Prover can decide to double-check its consistency too.

We do not need to provide these choices now, but we can plug them later easily if this method is failable.

Reply:
that makes sense to me! do we want to leave it as a no-op on the prover right now, or have the closure returnan `Option` and check on the prover if there is a PointOp

Reply:
Let's have both the method and the closure return `Result<(), VMError>`. Prover would do nothing (`Ok(())`) and verifier will batch.

Reply:
Oh, wait. Probably the closure will never fail. It just returns a PointOp.

Reply:
(edited)